Unify WASM crypto client with mobile

dani-garcia opened the pull request:
## 🎟️ Tracking

<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->

## 📔 Objective

With the removal of the lifetimes of the clients, we can remove the wasm-specific crypto client and instead use the one in bitwarden-core.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

dani-garcia committed:
Unify WASM crypto client with mobile

dani-garcia commented on crates/bitwarden-core/src/mobile/crypto_client.rs:
```diff
@@ -39,6 +43,24 @@ impl CryptoClient {
+    }
+}
+
+impl CryptoClient {
```
I've split the `impl CryptoClient` into two blocks, one with wasm_bindgen for the functions we're currently exporting to wasm, and one without it for the ones that we're not using.

We could also export all of them if we wanted, but to keep the PR simple I only include the ones actually in use.

dani-garcia commented on crates/bitwarden-core/Cargo.toml:
```diff
@@ -49,6 +50,7 @@ tsify-next = { workspace = true, optional = true }
 uniffi = { workspace = true, optional = true, features = ["tokio"] }
 uuid = { workspace = true }
 wasm-bindgen = { workspace = true, optional = true }
+wasm-bindgen-futures = { workspace = true, optional = true }
```
This dep is used internally for the wasm-bindgen macro as cryptoclient has some async functions.

dani-garcia committed:
Merge branch 'main' into ps/unify-wasm-crypto

# Conflicts:
#	Cargo.lock
#	crates/bitwarden-core/src/mobile/crypto.rs
#	crates/bitwarden-wasm-internal/src/client.rs
#	crates/bitwarden-wasm-internal/src/lib.rs

renovate[bot] committed:
[deps]: Update sonarsource/sonarqube-scan-action action to v5 (#291)

This PR contains the following updates:

| Package | Type | Update | Change |
|---|---|---|---|
|
[sonarsource/sonarqube-scan-action](https://redirect.github.com/sonarsource/sonarqube-scan-action)
| action | major | `v4.2.1` -> `v5.2.0` |

---

### Release Notes

<details>
<summary>sonarsource/sonarqube-scan-action
(sonarsource/sonarqube-scan-action)</summary>

###
[`v5.2.0`](https://redirect.github.com/SonarSource/sonarqube-scan-action/releases/tag/v5.2.0)

[Compare
Source](https://redirect.github.com/sonarsource/sonarqube-scan-action/compare/v5.1.0...v5.2.0)

##### What's Changed

- SQSCANGHA-90 remove mend dead conf by
[@&#8203;pierre-guillot-gh](https://redirect.github.com/pierre-guillot-gh)
in
[https://github.com/SonarSource/sonarqube-scan-action/pull/184](https://redirect.github.com/SonarSource/sonarqube-scan-action/pull/184)
- SQSCANGHA-89 Attempt to fix command injection by
[@&#8203;henryju](https://redirect.github.com/henryju) in
[https://github.com/SonarSource/sonarqube-scan-action/pull/186](https://redirect.github.com/SonarSource/sonarqube-scan-action/pull/186)
- SQSCANGHA-93 Fix madhead/semver-utils' version by
[@&#8203;csaba-feher-sonarsource](https://redirect.github.com/csaba-feher-sonarsource)
in
[https://github.com/SonarSource/sonarqube-scan-action/pull/187](https://redirect.github.com/SonarSource/sonarqube-scan-action/pull/187)
- SQSCANGHA-94 Update version update logic by
[@&#8203;csaba-feher-sonarsource](https://redirect.github.com/csaba-feher-sonarsource)
in
[https://github.com/SonarSource/sonarqube-scan-action/pull/188](https://redirect.github.com/SonarSource/sonarqube-scan-action/pull/188)
- SQSCANGHA-92 Validate scanner version by
[@&#8203;csaba-feher-sonarsource](https://redirect.github.com/csaba-feher-sonarsource)
in
[https://github.com/SonarSource/sonarqube-scan-action/pull/189](https://redirect.github.com/SonarSource/sonarqube-scan-action/pull/189)

**Full Changelog**:
https://github.com/SonarSource/sonarqube-scan-action/compare/v5...v5.2.0

###
[`v5.1.0`](https://redirect.github.com/SonarSource/sonarqube-scan-action/releases/tag/v5.1.0)

[Compare
Source](https://redirect.github.com/sonarsource/sonarqube-scan-action/compare/v5.0.0...v5.1.0)

#### What's Changed

- Update SonarScanner CLI to 7.1.0.4889 to support sonar.region=us by
[@&#8203;github-actions](https://redirect.github.com/github-actions) in
[https://github.com/SonarSource/sonarqube-scan-action/pull/183](https://redirect.github.com/SonarSource/sonarqube-scan-action/pull/183)

**Full Changelog**:
https://github.com/SonarSource/sonarqube-scan-action/compare/v5.0.0...v5.1.0

###
[`v5.0.0`](https://redirect.github.com/SonarSource/sonarqube-scan-action/releases/tag/v5.0.0)

[Compare
Source](https://redirect.github.com/sonarsource/sonarqube-scan-action/compare/v4.2.1...v5.0.0)

##### What's Changed

- SQSCANGHA-81 Update SonarScanner CLI to 7.0.2.4839 by
[@&#8203;github-actions](https://redirect.github.com/github-actions) in
[https://github.com/SonarSource/sonarqube-scan-action/pull/175](https://redirect.github.com/SonarSource/sonarqube-scan-action/pull/175)

**Full Changelog**:
https://github.com/SonarSource/sonarqube-scan-action/compare/v4...v5.0.0

</details>

---

### Configuration

📅 **Schedule**: Branch creation - "every 2nd week starting on the 2 week
of the year before 4am on Monday" (UTC), Automerge - At any time (no
schedule defined).

🚦 **Automerge**: Disabled by config. Please merge this manually once you
are satisfied.

♻ **Rebasing**: Whenever PR becomes conflicted, or you tick the
rebase/retry checkbox.

🔕 **Ignore**: Close this PR and you won't be reminded about this update
again.

---

- [ ] <!-- rebase-check -->If you want to rebase/retry this PR, check
this box

---

This PR was generated by [Mend Renovate](https://mend.io/renovate/).
View the [repository job
log](https://developer.mend.io/github/bitwarden/sdk-internal).

<!--renovate-debug:eyJjcmVhdGVkSW5WZXIiOiI0MC4xNi4wIiwidXBkYXRlZEluVmVyIjoiNDAuMTYuMCIsInRhcmdldEJyYW5jaCI6Im1haW4iLCJsYWJlbHMiOltdfQ==-->

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

Hinton commented on crates/bitwarden-core/src/mobile/crypto.rs:
**suggestion:** Could we lift this out of mobile?

dani-garcia replied:
Sounds reasonable yeah, do you think we should just lift it into the root? Or maybe into the key-management folder?

Hinton replied:
I think we can lift it into KM, @quexten would you be alright with that, I think everything currently falls under KM domain.

quexten replied:
Seems reasonable!

My question here (non-blocking, but needs clarification): Does key-management (the folder) really refer to the KM team, or the key-management capabilities (key store, key context) of the sdk?

The KM domain is not directly the same (Biometrics, Crypto Primitives).

dani-garcia replied:
So the way I'm thinking about this is like this:

The crypto client that I'm moving to the `key_management` folder is mostly a thin compatibility layer on top of the `bitwarden-crypto` types to make the functionality more usable for FFI (By exposing a client, instead of relying on types like `MasterKey`, etc, which aren't usable in FFI). In that sense, once KM officially owns `bitwarden-crypto`, this folder should come with, as it's fairly tied to the API exposed by `bitwarden-crypto`. In the future, if we find a way to expose the normal `bitwarden-crypto` API through FFI this module could be removed.

The other thing in the `key_management` folder is the definition of `KeyIds` used by the Key Store from `bitwarden-crypto`. I think of this and the Key Store as part of KM's field to expose safe to use crypto, so in my mind it would be part of your domain. The reason I defined these outside the crypto crate was to ensure the system was generic and easily expandable, but I'm not sure if that's worth it. If not, we can always move it into `bitwarden-crypto` directly and simplify ownership.

Hinton commented on crates/bitwarden-core/src/mobile/crypto.rs:
```diff
@@ -23,6 +25,7 @@ use crate::{
 
 /// Catch all error for mobile crypto operations.
 #[allow(missing_docs)]
+#[bitwarden_error(flat)]
```
The `MobileCryptoError` doesn't quite make sense if this is used from wasm.

dani-garcia replied:
I changed it to CryptoClientError now

dani-garcia committed:
Lift mobile crypto into key_management, and rename error

dani-garcia committed:
Make crypto functions private

Hinton committed:
Update InitUserCryptoRequest to use EncString (#298)

## 🎟️ Tracking

<!-- Paste the link to the Jira or GitHub issue or otherwise describe /
point to where this change is coming from. -->

## 📔 Objective

<!-- Describe what the purpose of this PR is, for example what bug
you're fixing or new feature you're adding. -->

This is an old type we created before we supported `EncString` as an FFI
type.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or
informed the documentation
  team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire!
-->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry
that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or
concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or
indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

dani-garcia committed:
Merge branch 'main' into ps/unify-wasm-crypto

dani-garcia committed:
Merge branch 'main' into ps/unify-wasm-crypto

# Conflicts:
#	crates/bitwarden-wasm-internal/src/client.rs
#	crates/bitwarden-wasm-internal/src/crypto.rs

dani-garcia committed:
Merge branch 'main' into ps/unify-wasm-crypto
